### DISCUSSION TITLE
Expose a ctrl callback in Unity Plugin?

### DISCUSSION THREAD

**Discussion by Balint-H** [google-deepmind/mujoco#233] [Ideas]
Hello, I'm really impressed with the state of the Unity Plugin! I've been having trouble with the stability of PD controllers even at moderate gains. Futhermore, I tried implementing my own update law.

I'm still building my understanding of the simulation loop, with help from the guidance on the Programming page of the docs. In the Unity plugin are the inputs of actuators through the `Control` property only sent to the simulation when `MjScene` calls `SyncUnityToMjState()`? If so, could the instability be related to that since then the control is updated after `mj_step()` has fully finished as opposed to in between of `mj_step1` and `mj_step2`?

I tried some ad-hoc changes in `MjScene.StepScene()` and added an `EventHandler` that C# scripts can hook into (called `ctrlCallback` below). `MjStepArgs` is an `EventArgs` that has the Model and Data pointers, with which the control laws can be evaluated.

Here's the edited method (with calls to the profiler omitted)

```
public unsafe void StepScene() {
    if (Model == null || Data == null) {
      throw new NullReferenceException("Failed to create Mujoco runtime.");
    }
    if (ctrlCallback != null){
      MujocoLib.mj_step1(Model, Data);
      ctrlCallback.Invoke(this, new MjStepArgs(Model, Data));
      MujocoLib.mj_step2(Model, Data);
    }
    else {
      MujocoLib.mj_step(Model, Data);
    }

    CheckForPhysicsException();
    SyncUnityToMjState();
  }
```

From early tests this seems to lead to more stable evaluation for custom PD control. Is this an appropriate approach to implement control laws with the Unity Plugin? Or am I misinterpreting/abusing the plugin?

I know something similar exists  with `mjcb_control`, but I couldn't find how to use it in the plugin.

**Comment by erez-tom** [comment]
Your fix is solid, and I can see how for some PD controllers the current off-by-1 design is an issue.  Another way of doing it would be to pull out the actuators from `_orderedComponents` into their own list and have them sync both before step2 and after (to pick up changes in `actuator_force`).

Using MuJoCo's own callbacks (e.g., `mjcb_control`) is also possible, but the current workflow is a hassle, see below.  We intend to revamp the callback mechanism after open-sourcing, and hopefully then we'll be able to offer a more flexible system.

The way it can be made to work right now: you write the callback in C/C++ and add an "assignment" function in this module that assigns the callback function pointer to the MuJoCo library's global callback variable (e.g., `mjcb_control`).  You then `DllImport` this assignment function into Unity, and the callback will be installed when you invoke it.
